hip/solver/common_trs_kernels.hip.hpp

```diff
@@ -0,0 +1,244 @@
+}
+
+
+template <typename ValueType, typename IndexType>
+void generate_kernel(std::shared_ptr<const HipExecutor> exec,
+                     const matrix::Csr<ValueType, IndexType> *matrix,
+                     solver::SolveStruct *solve_struct,
+                     const gko::size_type num_rhs, bool is_upper)
+{
+    auto hip_solve_struct =
+        reinterpret_cast<solver::hip::SolveStruct *>(solve_struct);
+    if (hipsparse::is_supported<ValueType, IndexType>::value) {
+        auto handle = exec->get_hipsparse_handle();
+        if (is_upper) {
+            GKO_ASSERT_NO_HIPSPARSE_ERRORS(hipsparseSetMatFillMode(
+                hip_solve_struct->factor_descr, HIPSPARSE_FILL_MODE_UPPER));
+        }
+
+        {
+            hipsparse::pointer_mode_guard pm_guard(handle);
+            hipsparse::csrsv2_buffer_size(
+                handle, HIPSPARSE_OPERATION_NON_TRANSPOSE,
+                matrix->get_size()[0], matrix->get_num_stored_elements(),
+                hip_solve_struct->factor_descr, matrix->get_const_values(),
+                matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(),
+                hip_solve_struct->solve_info,
+                &hip_solve_struct->factor_work_size);
+
+            // allocate workspace
+            if (hip_solve_struct->factor_work_vec != nullptr) {
+                exec->free(hip_solve_struct->factor_work_vec);
+            }
+            hip_solve_struct->factor_work_vec =
+                exec->alloc<void *>(hip_solve_struct->factor_work_size);
+
+            hipsparse::csrsv2_analysis(
+                handle, HIPSPARSE_OPERATION_NON_TRANSPOSE,
+                matrix->get_size()[0], matrix->get_num_stored_elements(),
+                hip_solve_struct->factor_descr, matrix->get_const_values(),
+                matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(),
+                hip_solve_struct->solve_info, hip_solve_struct->policy,
+                hip_solve_struct->factor_work_vec);
+        }
+
+    } else {
+        GKO_NOT_IMPLEMENTED;
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void solve_kernel(std::shared_ptr<const HipExecutor> exec,
+                  const matrix::Csr<ValueType, IndexType> *matrix,
+                  const solver::SolveStruct *solve_struct,
+                  matrix::Dense<ValueType> *trans_b,
+                  matrix::Dense<ValueType> *trans_x,
+                  const matrix::Dense<ValueType> *b,
+                  matrix::Dense<ValueType> *x)
+{
+    using vec = matrix::Dense<ValueType>;
+    auto hip_solve_struct =
+        reinterpret_cast<const solver::hip::SolveStruct *>(solve_struct);
+    if (hipsparse::is_supported<ValueType, IndexType>::value) {
+        ValueType one = 1.0;
+        auto handle = exec->get_hipsparse_handle();
+
+        {
+            hipsparse::pointer_mode_guard pm_guard(handle);
+            if (b->get_stride() == 1) {
+                auto temp_b = const_cast<ValueType *>(b->get_const_values());
```

Comment:
Why do we need this cast? Does hipSPARSE require non-const x?

Reply:
Good catch. No, we don't need that.
`csrsv2_solve`, `csrsm2_solve` use `const ValueType *`
`csrsm_solve` uses `ValueType *`